project/src/main/webapp/gmail.js

```diff
@@ -0,0 +1,365 @@
+            return false;
+        }
+        return true;
+    }
+
+    getEmailObjects() {
+        console.log(this.emailObjects);
+    }
+
+    /**
+     * Get Message with given ID.
+     *
+     * @param  {String} messageId ID of Message to get.
+     */
+    getMessage(messageId) {
+        const getMessagePromise = new Promise((resolve) => {
+            const messageRequest = gapi.client.gmail.users.messages.get({
+                'userId': 'me',
+                'id': messageId,
+                'format': "raw"
+            });
+            messageRequest.execute(response => {
+                //Convert from base64 encoding to text.
+                const decodedEmail = atob(response.raw.replace(/-/g, '+').replace(/_/g, '/'));
+                const emailBodyValue = this.getEmailBody(decodedEmail);
+                const emailDateValue = this.getEmailDate(decodedEmail);
+                const messageID = this.getMessageID(decodedEmail);
+                if (this.isActualEmail(emailBodyValue, emailDateValue)) {
+                    this.emailObjects[messageID] = { emailDate: emailDateValue, emailBody: emailBodyValue };
+                }
+                this.getEmailObjects();
+                resolve();
+            });
+        });
+        return getMessagePromise;
+    }
+
+    //Retrieve messages using hardcoded queries and the signed-in email.
+    listMessages() {
+        this.emailObjects = {};
+        const listMessagePromise = new Promise((resolve) => {
+            const promiseArray = [];
+            var getPageOfMessages = (request, result) => {
+                request.execute(resp => {
+                    result = result.concat(resp.messages);
+                    var nextPageToken = resp.nextPageToken;
+                    if (nextPageToken) {
+                        request = gapi.client.gmail.users.messages.list({
+                            'userId': 'me',
+                            'pageToken': nextPageToken,
+                            'q': this.question
+                        });
+                        getPageOfMessages(request, result);
+                    } else {
+                        for (var i = 0; i < result.length; i++) {
+                            if (result[i] === undefined) {
+                                console.log("There are no messages from the query");
+                            }
+                            else {
+                                console.log(result.length);
+                                console.log(result);
+                                console.log(result[i].id);
+                                // if (result[i].id === undefined) {
+                                //   this.getMessage(result[i].id);  
+                                // }
+                                promiseArray.push(this.getMessage(result[i].id));
+                            }
+                        }
+                        Promise.all(promiseArray).then(() => resolve()
```

Comment:
I just noticed a bug here, I don't think we've hit this recursive case yet, but if there is more than one page of search results we would.

The Promise.all(...).then(() => resolve()) needs to happen at the very end of this function and after any recursive calls, so can you try moving it after the outside call to getPageOfMessages? You can even simplify it into one promise:

```
    /** Retrieve messages using hardcoded queries and the signed-in email. */
    listMessages() {
        this.emailObjects = {};
        const promiseArray = [];
        let getPageOfMessages = (request, result) => {
            ...
        };
        getPageOfMessages(initialRequest, []);
        return Promise.all(promiseArray);
    }
```

Reply:
Done!